Bottom: cb13e87a22ec7057438542d7607ee64dc5e68834
Top:    aba0112e584ed6c1ee6b0499d276d2754adbbd78
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-24 01:10:23 +0100

Refresh of make-drawppv-always-draw-for-n

---

diff --git a/src/draw.c b/src/draw.c
index 2a097d9..46c569b 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -84,7 +84,7 @@ static const BoxType *clip_box = NULL;
 static void Redraw (bool, BoxTypePtr);
 static void DrawEverything (BoxTypePtr);
 static void DrawPPV (int group, const BoxType *);
-static int DrawLayerGroup (int, const BoxType *);
+static void DrawLayerGroup (int, const BoxType *);
 static void DrawPinOrViaLowLevel (PinTypePtr, bool);
 static void DrawPlainPin (PinTypePtr, bool);
 static void DrawPlainVia (PinTypePtr, bool);
@@ -312,11 +312,9 @@ static void
 PrintAssembly (const BoxType * drawn_area, int side_group, int swap_ident)
 {
   int save_swap = SWAP_IDENT;
-  int side = swap_ident ? SOLDER_LAYER : COMPONENT_LAYER;
 
   gui->set_draw_faded (Output.fgGC, 1);
   DrawLayerGroup (side_group, drawn_area);
-  DrawPPV (side_group, drawn_area);
   gui->set_draw_faded (Output.fgGC, 0);
 
   /* draw package */
@@ -381,25 +379,9 @@ DrawEverything (BoxTypePtr drawn_area)
       int group = drawn_groups[i];
 
       if (gui->set_layer (0, group, 0))
-	{
-	  if (DrawLayerGroup (group, drawn_area) && !gui->gui)
-	    {
-	      r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
-	      r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
-	      /* draw element pads */
-	      if (group == component || group == solder)
-		{
-                  side = (group == solder) ? SOLDER_LAYER : COMPONENT_LAYER;
-		  r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
-		}
-
-	      /* draw holes */
-	      plated = -1;
-	      r_search (PCB->Data->pin_tree, drawn_area, NULL, hole_callback, &plated);
-	      r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, &plated);
-	    }
-	}
+        DrawLayerGroup (group, drawn_area);
     }
+
   if (TEST_FLAG (CHECKPLANESFLAG, PCB) && gui->gui)
     return;
 
@@ -567,7 +549,8 @@ pad_callback (const BoxType * b, void *cl)
 }
 
 /* ---------------------------------------------------------------------------
- * draws pins pads and vias
+ * Draws pins pads and vias - Always draws for non-gui HIDs,
+ * otherwise drawing depends on PCB->PinOn and PCB->ViaOn
  */
 static void
 DrawPPV (int group, const BoxType * screen)
@@ -575,7 +558,7 @@ DrawPPV (int group, const BoxType * screen)
   int component_group = GetLayerGroupNumberByNumber (component_silk_layer);
   int solder_group = GetLayerGroupNumberByNumber (solder_silk_layer);
 
-  if (PCB->PinOn || doing_assy)
+  if (PCB->PinOn || !gui->gui)
     {
       /* draw element pins */
       r_search (PCB->Data->pin_tree, screen, NULL, pin_callback, NULL);
@@ -595,7 +578,7 @@ DrawPPV (int group, const BoxType * screen)
     }
 
   /* draw vias */
-  if (PCB->ViaOn || doing_assy)
+  if (PCB->ViaOn || !gui->gui)
     {
       r_search (PCB->Data->via_tree, screen, NULL, via_callback, NULL);
       r_search (PCB->Data->via_tree, screen, NULL, hole_callback, NULL);
@@ -811,16 +794,20 @@ DrawLayer (LayerTypePtr Layer, const BoxType * screen)
  * draws one layer group.  Returns non-zero if pins and pads should be
  * drawn with this group.
  */
-static int
-DrawLayerGroup (int group, const BoxType * screen)
+static void
+DrawLayerGroup (int group, const BoxType * drawn_area)
 {
   int i, rv = 1;
   int layernum;
+  int plated;
+  int side;
   LayerTypePtr Layer;
   int n_entries = PCB->LayerGroups.Number[group];
   Cardinal *layers = PCB->LayerGroups.Entries[group];
+  int component_group = GetLayerGroupNumberByNumber (component_silk_layer);
+  int solder_group = GetLayerGroupNumberByNumber (solder_silk_layer);
 
-  clip_box = screen;
+  clip_box = drawn_area;
   for (i = n_entries - 1; i >= 0; i--)
     {
       layernum = layers[i];
@@ -829,11 +816,13 @@ DrawLayerGroup (int group, const BoxType * screen)
 	  strcmp (Layer->Name, "route") == 0)
 	rv = 0;
       if (layernum < max_copper_layer && Layer->On)
-	DrawLayerCommon (Layer, screen, true);
+	DrawLayerCommon (Layer, drawn_area, true);
     }
   if (n_entries > 1)
     rv = 1;
-  return rv;
+
+  if (rv == 1 && (!gui->gui || doing_assy))
+    DrawPPV (group, drawn_area);
 }
 
 /* ---------------------------------------------------------------------------
